libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on cpufreq(8) from BPF-Perf-Tools-Book by Brendan Gregg.
+// 10-OCT-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <linux/perf_event.h>
+#include <asm/unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "cpufreq.h"
+#include "cpufreq.skel.h"
+#include "trace_helpers.h"
+
+#define FREQ	99
+
+static struct env {
+	int duration;
+	bool verbose;
+} env = {
+	.duration = -1
+};
+
+const char *argp_program_version = "cpufreq 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Sampling CPU freq system-wide & by process. Ctrl-C to end.\n"
+"\n"
+"USAGE: cpufreq [--help] [-d DURATION]\n"
+"\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Duration to trace in seconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		errno = 0;
+		env.duration = strtol(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int nr_cpus;
+
+static int open_and_attach_perf_event(int freq, struct bpf_program *prog,
+				struct bpf_link *links[])
+{
+	struct perf_event_attr attr = {
+		.type = PERF_TYPE_SOFTWARE,
+		.freq = 1,
+		.sample_period = freq,
+		.config = PERF_COUNT_SW_CPU_CLOCK,
+	};
+	int i, fd;
+
+	for (i = 0; i < nr_cpus; i++) {
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int init_freqs_hmz(__u32 *freqs_mhz, int nr_cpus)
```

Comment:
have you tried reading /sys/devices/system/cpu/cpu<N>/cpufreq/cpuinfo_cur_freq instead? Seems much simpler to parse (and should be more performant, of course due to simplicity). Unless there is some reason to not use it (e.g., it's not available on some configurations, etc).

Reply:
I considered that, but I found some environments may do not have  `cpufreq`  in `sysfs` (e.g., wsl2, kvm),  so I chose `cpuinfo` instead.

Reply:
Fixed by parsing `sysfs`, thanks!